lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -166,6 +173,118 @@ static bool isTypeTooComplex(Type t) {
   return TypeWidth >= TypeWidthThreshold || TypeDepth >= TypeDepthThreshold;
 }
 
+/// Check whether the type T1 is contained in the type T2
+static bool isContainedIn(CanType T1, CanType T2) {
+  return T2.findIf(
+      [&T1](Type T) -> bool { return T->getCanonicalType() == T1; });
+}
+
+/// Checks if a second substitution list is a expanded version of
+/// the first substitution list.
+/// This is the case if at least one of the substitution type in Subs2 is
+/// "bigger" than the corresponding substitution type in Subs1.
+/// Type T2 is "smaller" than type T1 if T2 is structurally contained in T1.
+static bool growingSubstitutions(SubstitutionList Subs1,
```

Comment:
I think recursive constraints allow this to be "fooled":

```swift
protocol P {
    associatedtype X: P
}

struct Start {}
struct Step<Param> {}

struct Outer<Param>: P {
    typealias X = Outer<Step<Param>>
}

func foo<T: P>(_: T.Type) {
    foo(T.X.self)
}
foo(Outer<Start>.self)
```

The `T` in `foo` looks like:

- `Outer<Start>`
- `Outer<Step<Start>>`
- `Outer<Step<Step<Start>>>`
- ...

None of which are considered growing because they're never directly contained within one another.

Maybe first pattern-matching through any common prefix (I think it's `TypeMatcher` that does this?) and only then checking for growing-ness of the mismatched sections is enough. Combined with doing this at every level of the call stack, this seems like it should even catch things with arbitrary recursion depth like (variations of):

```swift
struct Outer1<Param>: P {
    typealias X = Outer2<Param>
}
struct Outer2<Param>: P {
    typealias X = Outer3<Param>
}
// ...
struct OuterN<Param>: P {
    typealias X = Outer1<Step<Param>>
}

foo(Outer1<Start>.self)
```